src/mbgl/tile/tile_cache.cpp

```diff
@@ -26,43 +41,54 @@ void TileCache::setSize(size_t size_) {
 }
 
 namespace {
-
-/// This exists solely to prevent a problem where temporary lambda captures
-/// are retained for the duration of the scope instead of being destroyed immediately.
-template <typename T>
+// Capturing `std::vector<std::unique_ptr<Tile>>` directly produces an error related to
+// copying, but this somehow avoids the same problem.  It may be possible to eliminate it.
 struct CaptureWrapper {
-    CaptureWrapper(std::unique_ptr<T>&& item_)
-        : item(std::move(item_)) {}
-    CaptureWrapper(const CaptureWrapper& other)
-        : item(other.item) {}
-    std::shared_ptr<T> item;
+    std::vector<std::unique_ptr<Tile>> releases;
+
+    CaptureWrapper(std::vector<std::unique_ptr<Tile>>&& x)
+        : releases(std::move(x)) {}
+    CaptureWrapper(CaptureWrapper&&) = default;
+    CaptureWrapper& operator=(CaptureWrapper&&) = default;
+    CaptureWrapper(CaptureWrapper const&) = delete;
+    CaptureWrapper& operator=(CaptureWrapper const&) = delete;
 };
 } // namespace
 
 void TileCache::deferredRelease(std::unique_ptr<Tile>&& tile) {
     MLN_TRACE_FUNC();
 
     tile->cancel();
+    pendingReleases.push_back(std::move(tile));
+}
 
-    // The `std::function` must be created in a separate statement from the `schedule` call.
-    // Creating a `std::function` from a lambda involves a copy, which is why we must use
-    // `shared_ptr` rather than `unique_ptr` for the capture.  As a result, a temporary holds
-    // a reference until the construction is complete and the lambda is destroyed.
-    // If this temporary outlives the `schedule` call, and the function is executed immediately
-    // by a waiting thread and is already complete, that temporary reference ends up being the
-    // last one and the destruction actually occurs here on this thread.
-    std::function<void()> func{[tile_{CaptureWrapper<Tile>{std::move(tile)}}, this]() mutable {
-        tile_.item = {};
-
-        std::lock_guard<std::mutex> counterLock(deferredSignalLock);
-        deferredDeletionsPending--;
-        deferredSignal.notify_all();
-    }};
+void TileCache::deferPendingReleases() {
+    MLN_TRACE_FUNC();
 
-    std::unique_lock<std::mutex> counterLock(deferredSignalLock);
-    deferredDeletionsPending++;
+    constexpr std::size_t scheduleThreshold = 1;
+    if (pendingReleases.size() < scheduleThreshold) {
+        return;
+    }
+
+    // Block destruction until the cleanup task is complete
+    {
+        std::lock_guard counterLock{deferredSignalLock};
+        deferredDeletionsPending++;
+    }
+
+    threadPool.schedule(
+        util::unique_function<void()>{[this, wrap_{CaptureWrapper{std::move(pendingReleases)}}]() mutable {
```

Comment:
You are moving the data member `pendingReleases` here.

Reply:
Interesting, there used to be language on cppreference implying that was valid, but it's gone now.  Seems like it was removed since 2022, based on [this](https://news.ycombinator.com/item?id=32907069).

Reply:
<img width="659" alt="image" src="https://github.com/user-attachments/assets/2b533993-5a56-4e51-82a7-340d5a5ae186">
<img width="673" alt="image" src="https://github.com/user-attachments/assets/b509f10f-5280-4be1-9006-b3689361f869">

Reply:
I changed it to move the elements to a temporary yesterday, but it sounds like it's sufficient to use `clear()` to return it to a well-defined valid state, right?